File: C++/Grind75/Week-3/3.longest-substring-without-repeating-characters.cpp
#include <bits/stdc++.h>
using namespace std;
/*
 * @lc app=leetcode id=3 lang=cpp
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
class Solution
{
public:
    int lengthOfLongestSubstring(string s)
    {
        unordered_map<char, int> cmp;
        int l = 0, r = 0, ans = 0;
        while (r < s.size())
        {
            if (cmp.find(s[r]) != cmp.end())
            {
                l = max(l, cmp[s[r]] + 1);
            }
            cmp[s[r]] = r;
            ans = max(ans, r - l + 1);
            r++;
        }

        return ans;
    }
};
// @lc code=end

File: Java/Grind75/Week 3/3.longest-substring-without-repeating-characters.java
import java.util.*;

/*
 * @lc app=leetcode id=3 lang=java
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> set = new HashSet<>();
        int max = 0, l = 0, r = 0, length = s.length();
        while (r < length) {
            while (set.contains(s.charAt(r))) {
                set.remove(s.charAt(l));
                l++;
            }
            set.add(s.charAt(r));
            r++;
            max = Math.max(max, r - l);
        }
        return max;
    }
}
// @lc code=end
